File: app/src/main/java/com/example/dkalev/remember/flashcard/CardRecyclerTouchListener.java
package com.example.dkalev.remember.flashcard;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.GestureDetector;
import android.view.MotionEvent;

import static android.view.MotionEvent.INVALID_POINTER_ID;

public class CardRecyclerTouchListener implements RecyclerView.OnItemTouchListener  {

    private GestureDetector mGestureDetector;
    private ClickListener mClickListener;

    private float mLastTouchX;
    private float mLastTouchY;
    private int mActivePointerId = INVALID_POINTER_ID;
    public static int SWIPE_OUT = 0;
    public static int SWIPE_IN = 1;


    public CardRecyclerTouchListener(Context context, final RecyclerView rv, final ClickListener clickListener){
        mClickListener = clickListener;
        mGestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener(){
            @Override
            public boolean onSingleTapUp(MotionEvent e) {
                FlashcardView child = (FlashcardView)rv.findChildViewUnder(e.getX(), e.getY());
                mClickListener.onClick(child);
                return true;
            }

            @Override
            public void onLongPress(MotionEvent e) {
                FlashcardView child = (FlashcardView) rv.findChildViewUnder(e.getX(), e.getY());

                if (child != null && clickListener != null) {
                    clickListener.onLongClick(child);
                }
            }
        });
    }

    @Override
    public boolean onInterceptTouchEvent(RecyclerView rv, MotionEvent e) {

        FlashcardView v = (FlashcardView) rv.findChildViewUnder(e.getX(), e.getY());
        mGestureDetector.onTouchEvent(e);
        if (!v.isFlipped()) {
            return false;
        }

        final int action = e.getActionMasked();

        switch (action) {
            case MotionEvent.ACTION_DOWN: {
                final int pointerIndex = e.getActionIndex();
                final float x = e.getX(pointerIndex);
                final float y = e.getY(pointerIndex);

                // Remember where we started (for dragging)
                mLastTouchX = x;
                mLastTouchY = y;
                // Save the ID of this pointer (for dragging)
                mActivePointerId = e.getPointerId(0);
                break;
            }

            case MotionEvent.ACTION_MOVE: {
                // Find the index of the active pointer and fetch its position
                final int pointerIndex =
                        e.findPointerIndex(mActivePointerId);

                final float x = e.getX(pointerIndex);
                final float y = e.getY(pointerIndex);

                // Calculate the distance moved
                final float dx = x - mLastTouchX;
                final float dy = y - mLastTouchY;

                v.setTranslationX(v.getTranslationX() + dx);
                v.setTranslationY(v.getTranslationY() + dy);

                // Remember this touch position for the next move event
                mLastTouchX = x;
                mLastTouchY = y;

                break;
            }

            case MotionEvent.ACTION_UP: {
                mActivePointerId = INVALID_POINTER_ID;

                if (v.getTranslationX() < -v.getWidth()/4 || v.getTranslationX() > v.getWidth()/4
                        || v.getTranslationY() < - v.getHeight()/4 || v.getTranslationY() > v.getHeight()/4){
                    mClickListener.onSwipe(v, SWIPE_OUT);
                }else{
                    mClickListener.onSwipe(v, SWIPE_IN);
                }

                break;
            }

            case MotionEvent.ACTION_CANCEL: {
                mActivePointerId = INVALID_POINTER_ID;
                break;
            }

            case MotionEvent.ACTION_POINTER_UP: {

                final int pointerIndex = e.getActionIndex();
                final int pointerId = e.getPointerId(pointerIndex);

                if (pointerId == mActivePointerId) {
                    // This was our active pointer going up. Choose a new
                    // active pointer and adjust accordingly.
                    final int newPointerIndex = pointerIndex == 0 ? 1 : 0;
                    mLastTouchX = e.getX(newPointerIndex);
                    mLastTouchY = e.getY(newPointerIndex);
                    mActivePointerId = e.getPointerId(newPointerIndex);
                }
                break;
            }
        }
        return false;
    }

    @Override
    public void onTouchEvent(RecyclerView rv, MotionEvent e) {

    }

    @Override
    public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {

    }

    public interface ClickListener {
        void onClick(FlashcardView view);

        void onLongClick(FlashcardView view);

        void onSwipe(FlashcardView view, int swipe_type);
    }
}

File: app/src/main/java/com/example/dkalev/remember/ui/card/CardRecyclerTouchListener.kt
package com.example.dkalev.remember.ui.card

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.MotionEvent.INVALID_POINTER_ID

class CardRecyclerTouchListener(context: Context, rv: RecyclerView, private val clickListener: ClickListener?) : RecyclerView.OnItemTouchListener {

    private val mGestureDetector: GestureDetector

    private var lastTouchX: Float = 0.toFloat()
    private var lastTouchY: Float = 0.toFloat()
    private var activePointerId = INVALID_POINTER_ID
    private lateinit var cardView: FlashcardView
    private var isSwiping: Boolean = false
    private var viewInitialized: Boolean = false


    init {
        mGestureDetector = GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
            override fun onSingleTapUp(e: MotionEvent): Boolean {
                val child = rv.findChildViewUnder(e.x, e.y) as FlashcardView
                clickListener!!.onClick(child)
                return true
            }

            override fun onLongPress(e: MotionEvent) {
                val child = rv.findChildViewUnder(e.x, e.y) as FlashcardView

                clickListener?.onLongClick(child)
            }
        })
    }

    override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
        mGestureDetector.onTouchEvent(e)

        if (!viewInitialized) {
            cardView = rv.findChildViewUnder(e.x, e.y) as FlashcardView
            viewInitialized = true
        }

        if (!cardView.isFlipped) {
            return true
        }

        val action = e.actionMasked
        if (action == MotionEvent.ACTION_DOWN){
            val pointerIndex = e.actionIndex
            val x = e.getX(pointerIndex)
            val y = e.getY(pointerIndex)

            // Remember where we started (for dragging)
            lastTouchX = x
            lastTouchY = y
            // Save the ID of this pointer (for dragging)
            activePointerId = e.getPointerId(0)

            isSwiping = true
        }
        return true
    }

    override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {
        mGestureDetector.onTouchEvent(e)
        if (isSwiping){

            val action = e.actionMasked

            when (action) {
                MotionEvent.ACTION_MOVE -> {
                    // Find the index of the active pointer and fetch its position
                    val pointerIndex = e.findPointerIndex(activePointerId)

                    val x = e.getX(pointerIndex)
                    val y = e.getY(pointerIndex)

                    // Calculate the distance moved
                    val dx = x - lastTouchX
                    val dy = y - lastTouchY

                    cardView.translationX = cardView.translationX + dx
                    cardView.translationY = cardView.translationY + dy

                    // Remember this touch position for the next move event
                    lastTouchX = x
                    lastTouchY = y
                }

                MotionEvent.ACTION_UP -> {
                    activePointerId = INVALID_POINTER_ID
                    isSwiping = false
                    viewInitialized = false

                    if (cardView!!.translationX < -cardView.width / 4 || cardView.translationX > cardView.width / 4
                            || cardView.translationY < -cardView.height / 4 || cardView.translationY > cardView.height / 4) {
                        clickListener!!.onSwipe(cardView, SWIPE_OUT)
                    } else {
                        clickListener!!.onSwipe(cardView, SWIPE_IN)
                    }
                }

                MotionEvent.ACTION_CANCEL -> {
                    activePointerId = INVALID_POINTER_ID
                    isSwiping = false
                    viewInitialized = false
                }

                MotionEvent.ACTION_POINTER_UP -> {

                    val pointerIndex = e.actionIndex
                    val pointerId = e.getPointerId(pointerIndex)

                    if (pointerId == activePointerId) {
                        // This was our active pointer going up. Choose a new
                        // active pointer and adjust accordingly.
                        val newPointerIndex = if (pointerIndex == 0) 1 else 0
                        lastTouchX = e.getX(newPointerIndex)
                        lastTouchY = e.getY(newPointerIndex)
                        activePointerId = e.getPointerId(newPointerIndex)
                    }
                }
            }
        }
    }

    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {

    }

    interface ClickListener {
        fun onClick(view: FlashcardView)

        fun onLongClick(view: FlashcardView?)

        fun onSwipe(view: FlashcardView, swipe_type: Int)
    }

    companion object {
        var SWIPE_OUT = 0
        var SWIPE_IN = 1
    }
}